### ISSUE TITLE
`accepts` is never called on iOS 8.4.1

### ISSUE THREAD

**Issue by Bushjumper** [bevacqua/dragula#164]
We're using Dragula JS in a hybrid mobile app. We noticed that after upgrading to the latest iOS version earlier (8.4.1), dragging doesn't work as it previously did.

Dragging an item starts ok but the 'accepts' callback is never triggered because getElementBehindPoint() seems to always return the whole document, not the specific element behind the point. This means that dropping isn't possible.

Of course, this is probably more of an iOS issue than a Dragula one, but I thought it worth mentioning here in case others have the same problem, or there may be a workaround or known fix?

**Comment by bevacqua** [comment]
I have a few action items for you, if you don't mind.

- Could this be related? http://stackoverflow.com/questions/19007553/document-elementfrompointx-y-pure-javascript-alternative

- Could you log the values of `x` and `y`? Maybe those are the crux of the problem, and we're getting clientX/clientY wrong for iOS 8.4.1

- Otherwise we probably should look into adding a case where if the document body is returned we make one more attempt at getting the element, but manually, looping over the DOM somehow.

**Comment by bevacqua** [comment]
I couldn't reproduce this issue on the demo page.

**Comment by fluff** [comment]
Having a similar issue in an in-house app (we are using Bootstrap3 and jQuery 1.11). 

We are using a "moves:" function to restrict drag-starts to a particular class of drag-handles, but we don't have any custom "accepts:" function. Dragging starts fine, but dragula doesn't seem to find any valid drop points, so .on('drop') never triggers and whatever is being dragged just pops back to its original position.

The iOS simulator (which is on version "8.4.0", there are no 8.4.1 updates for the simulator) works on both the demo page and in our app.

A real iPhone6+ running iOS 8.4.1 works on the demo page, but not in our app.

**Comment by fluff** [comment]
For us, the root cause seemed to be the following bootstrap style:
```css
.col-xs-1, .col-sm-1, .col-md-1, .col-lg-1, .col-xs-2, .col-sm-2, .col-md-2, .col-lg-2, .col-xs-3, .col-sm-3, .col-md-3, .col-lg-3, .col-xs-4, .col-sm-4, .col-md-4, .col-lg-4, .col-xs-5, .col-sm-5, .col-md-5, .col-lg-5, .col-xs-6, .col-sm-6, .col-md-6, .col-lg-6, .col-xs-7, .col-sm-7, .col-md-7, .col-lg-7, .col-xs-8, .col-sm-8, .col-md-8, .col-lg-8, .col-xs-9, .col-sm-9, .col-md-9, .col-lg-9, .col-xs-10, .col-sm-10, .col-md-10, .col-lg-10, .col-xs-11, .col-sm-11, .col-md-11, .col-lg-11, .col-xs-12, .col-sm-12, .col-md-12, .col-lg-12 {
  position: relative;
}
```
Our workaround was to add: ```style="position:static"``` to the containing ```<div class="col-md-6">```

**Comment by bevacqua** [comment]
Yeah in general you want not to mess with styles. I'll be closing this one.

**Comment by fluff** [comment]
Not sure how this is "messing with styles", I was just trying to apply dragula on some table <tr> rows which happened to be hosted deep inside a common bootstrap-css column. :) 

Anyways just wanted to leave a note about this workaround, in case anyone else gets bitten by it. 

It's pretty strange how this seems to only affect iOS 8.4.1. iOS 8.4.0 in the simulator and iOS 9.0 beta on another device didn't exhibit the problem.

**Comment by bevacqua** [comment]
I meant using bootstrap in general. :-)  
Where is it that you're placing `position: static` to fix the issue?  
Do you think that should be a part of `dragula`'s styles?

**Comment by fluff** [comment]
I put the position:static on the containing ```<div class="col-md-6">```, turning it into ```<div class="col-md-6" style="position:static">``` (Far "outside" the stuff that I was applying dragula to). 

I'm not sure why this is necessary. Maybe there's a bug in Mobile Safari in iOS 8.4.1 in the way "position:relative" affects how nodes inside it are hit-matched for "drop zones". Honestly I was just deleting stuff out of bootstrap.css until dragula started working :)

Sorry I haven't had the time to create minimal re-pro :-/

**Comment by bevacqua** [comment]
I presume that the issue stems from the way mirror position is calculated?

**Comment by Bushjumper** [comment]
Thanks @fluff - that helps us too. We're looking into possible conflicting styles, no workaround as yet but it's another path to try. Agreed that the iOS 8.4.1 thing is pretty wonky...

@bevacqua - to get back to you about your action items: it seems that the x and y values are still being logged correctly but the whole html document is always being returned rather than the element directly behind the dragging item. 

We'll keep digging. Thanks guys

**Comment by bevacqua** [comment]
So `elementFromPoint` is the culprit? That'd be hard to fix :<

**Comment by fluff** [comment]
I'm not sure. I tried hooking the phone up to the safari debugger, and set some breakpoints looking at clientX/clientY positions, but it was a bit tricky to make sense of the stuff, so I gave up trying to understand the root cause when I found the quick workaround. Personally I don't mind if you keep the issue closed, just wanted to add my 2cents :) Maybe iOS 9.0 will fix it (we couldn't repro on a phone running the beta).

**Comment by Bushjumper** [comment]
Yip, in our case, it seems to be document.elementFromPoint() that not returning what dragula expects - but only on iOS 8.4.1!

**Comment by bevacqua** [comment]
I'll reopen, we should try and find a fix for the underlying root cause. How about if we went up the DOM chain and added `static !important` to all the things, then calculated `elementFromPoint`, then restored/removed the style?

Pseudo code

```js
el = target;
while (el.parentElement) {
  el.style.position = 'static !important';
  el = el.parentElement;
}
var behind = document.elementFromPoint(x,y);
el = target;
while (el.parentElement) {
  el.style.position = 'whatever'; // (restore to inline style value)
  el = el.parentElement;
}
```

It might be too slow though.

**Comment by xStrom** [comment]
I can confirm that ```document.elementFromPoint()``` seems to be broken in iOS 8.4.1. I found this thread here while debugging an unrelated project.

**Comment by xStrom** [comment]
[I fixed this issue for another drag & drop library](https://github.com/timruffles/ios-html5-drag-drop-shim/pull/47). Maybe it contains useful info for you guys.

Basically what I found was that if I set ```display: none;``` to an element that is covered by the coordinates given to ```elementFromPoint```, then even if that element has ```pointer-events: none;```, Safari gets confused. When I added a 200ms delay between changing ```display``` and calling ```elementFromPoint```, it started occasionally working, but not always. I ended up removing the ```display``` modifying code completely and now ```elementFromPoint``` seems to work as it should.

**Comment by bevacqua** [comment]
Awesome! Thanks for the pointers

**Comment by Bushjumper** [comment]
FYI: I've just upgraded to iOS 9 and the issue seems to be resolved

**Comment by rwillbrown** [comment]
Something appears to have regressed. I ended up here after noticing drag-and-drop wasn't working on iOS in my project; clientX/clientY are undefined. I checked the demo page as well and got the same result. Tested with iOS 9.0.2 on device and 9.0 in the simulator.

**Comment by bevacqua** [comment]
Can you please identify what version introduced the regression?

**Comment by rwillbrown** [comment]
It appears to be v3.3.1. It attempts to use clientX/Y to prevent the drag from starting if the mouse position has no delta, and since it's undefined, all drags get ignored.

I was able to work around it quickly by changing the condition on line 158 to make sure there are X/Y positions to compare to before dropping out: (_moveX && _moveY && e.clientX === _moveX && e.clientY === _moveY).

**Comment by bevacqua** [comment]
Can you PR?

**Comment by rwillbrown** [comment]
Submitted #228.
I had to modify some of the other checks for mouse buttons as well, they were also preventing all touch drags.

**Comment by bevacqua** [comment]
Resolved in `3.5.3`
